### PR TITLE
Make it possible to adjust simulator scale factor

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
These changes are based on https://github.com/appium/appium/issues/6925

### CODE REVIEW

File: lib/simulator-xcode-6.js [hunk 1]
```diff
@@ -47,6 +49,17 @@ class SimulatorXcode6 {
     this.extraStartupTime = EXTRA_STARTUP_TIME;
   }
 
+  setScaleFactor (newScaleFactor) {
+    let supportedScales = ['1.0', '0.75', '0.5', '0.33', '0.25'];
+    if (supportedScales.indexOf(newScaleFactor) < 0) {
+      let msg = 'Only "' + supportedScales + '" values are supported as scale factors. "' +
```
Comment (jlipps) [inline lib/simulator-xcode-6.js]: we use template strings now instead of concatenation

File: lib/simulator-xcode-6.js [hunk 2]
```diff
@@ -47,6 +49,17 @@ class SimulatorXcode6 {
...
+                newScaleFactor + '" is passed instead.';
+      log.debug(msg);
```
Comment (jlipps) [inline lib/simulator-xcode-6.js]: for this pattern we have a method: `log.errorAndThrow(msg);`

File: lib/simulator-xcode-6.js [hunk 3]
```diff
@@ -239,6 +252,12 @@ class SimulatorXcode6 {
     // start simulator
     let simulatorApp = path.resolve(await getXcodePath(), 'Applications', this.simulatorApp);
     let args = [simulatorApp, '--args', '-CurrentDeviceUDID', this.udid];
+    if (!!this.scaleFactor) {
+      let stat = await this.stat();
+      let additionalArgs = ['-SimulatorWindowLastScale-com.apple.CoreSimulator.SimDeviceType.' +
```
Comment (jlipps) [inline lib/simulator-xcode-6.js]: template strings here as well

**Comment by imurchie** [comment]
Looks good to me! I agree with the comments of @jlipps and one quick run of `gulp eslint`.

File: lib/simulator-xcode-6.js (hunk 1 above)
Comment (mykola-mokhnach) [inline lib/simulator-xcode-6.js]: Done. Finally, something useful has been added to the ECMAScript standard %)

File: lib/simulator-xcode-6.js (hunk 2 above)
Comment (mykola-mokhnach) [inline lib/simulator-xcode-6.js]: Fixed

File: lib/simulator-xcode-6.js (hunk 3 above)
Comment (mykola-mokhnach) [inline lib/simulator-xcode-6.js]: fixed

**Comment by mykola-mokhnach** [comment]
@jlipps I have one question there about template strings usage - is it safe in general? [Here](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/template_strings), for example, it is said, that template strings SHOULD NOT contain any user input,  since the engine will interpret any valid JS expression inside

```
${...} 
```

specifier.

**Comment by imurchie** [comment]
If it were bare user input there would be the potential for a problem, but in these cases it is from within Appium. The first (logging the scale) will have a value that has been vetted in the driver (through validation on the desired capabilities) and the second is a computed value.

**Comment by imurchie** [comment]
Linting is still broken. Can you re-install dependencies? Pre-commit hooks are added so it won't depend on CI to check. You can also run `gulp eslint`.

```
rm -rf node_modules
npm install
```

**Comment by mykola-mokhnach** [comment]
@imurchie It's happy now :)

**Comment by mykola-mokhnach** [comment]
@jlipps anything else to do to have this PR merged?

**Comment by imurchie** [comment]
@mykola-mokhnach Can you squash this into a single commit? If you rebase with master the tests will (hopefully) be less flakey on Travis, too. Thanks! Then we can get this in!

**Comment by mykola-mokhnach** [comment]
this CLA system seems to be broken there - I've signed the agreement, but it still shows it is not signed. I'm tired playing with robots. Closing PR.

screen

**Comment by imurchie** [comment]
Weird. Yes, something is wrong there.

But there is no need to stop the PR. We can merge things even with that alert on.

**Comment by JSFOwner** [comment]
Sorry for the issue you are experiencing with the CLA check. I can verify that the user signed the CLA so this is ok to merge. Here is the entry in our database

`mykola-mokhnach Nov 1, 2016 2:36:00 AM`

For some reason the necessary author information is missing from the commits when pulled via the API in order to do the CLA signature comparison. The working theory is that due to the branch being deleted, that info is no longer available but this is just a theory. If that branch can be restored, let us know and we can rerun the check from this end and it should hopefully straighten out.

**Comment by mykola-mokhnach** [comment]
@JSFOwner I've restored the branch, but it still does not work as expected and shows me that LA has not been signed yet.

**Comment by jlipps** [comment]
@mykola-mokhnach you need to follow the instructions at https://help.github.com/articles/setting-your-email-in-git/ to ensure your git client is correctly linked with your github user. That should satisfy the CLA bot, and is something we require in any case.

File: lib/simulator-xcode-6.js [hunk 4]
```diff
@@ -239,6 +250,12 @@ class SimulatorXcode6 {
     // start simulator
     let simulatorApp = path.resolve(await getXcodePath(), 'Applications', this.simulatorApp);
     let args = [simulatorApp, '--args', '-CurrentDeviceUDID', this.udid];
+    if (this.scaleFactor) {
+      let stat = await this.stat();
+      let formattedDeviceName = stat.name.replace(/\s+/g, '-');
+      let argumentName = `-SimulatorWindowLastScale-com.apple.CoreSimulator.SimDeviceType.${formattedDeviceName}`;
+      args.push.apply(args, [argumentName, this.scaleFactor]);
```
Comment (jlipps) [inline lib/simulator-xcode-6.js]: why not just:

```js
args.push(argumentName, this.scaleFactor);
```

Comment (mykola-mokhnach) [inline lib/simulator-xcode-6.js]: fixed

**Comment by mykola-mokhnach** [comment]
@jlipps fixed

**Comment by imurchie** [comment]
Looks good to me.

**Comment by imurchie** [comment]
If you can squash into a single commit, we can get this going!

**Comment by jlipps** [comment]
i squashed and merged

**Comment by jlipps** [comment]
and published in 1.8.0

**Pull request merged**